Sem. 03/Pract. 03/mySolutions/Task01.cpp

```diff
@@ -0,0 +1,166 @@
+#include <iostream>
+#include <fstream>
+
+const size_t MAX_ADDRESS_LENGTH = 26;
+const char* FILE_PATH = "orders.bin";
+
+struct PizzaOrder
+{
+	double price;
+	char address[MAX_ADDRESS_LENGTH];
+};
+
+size_t getFileSize(std::ifstream& in)
+{
+	size_t currentPos = in.tellg();
+	in.seekg(0, std::ios::end);
+	size_t fileSize = in.tellg();
+	in.seekg(currentPos);
+	return fileSize;
+}
+
+size_t strLength(const char* str)
+{
+	size_t length = 0;
+
+	while (*str != 0)
+	{
+		str++;
+		length++;
+	}
+
+	return length;
+}
+
+int strCompare(const char* str1, const char* str2)
+{
+	size_t str1Len = strLength(str1);
+	size_t str2Len = strLength(str2);
+
+	size_t lenToUse = str1Len >= str2Len ? str1Len : str2Len;
+
+	for (size_t i = 0; i < lenToUse; i++)
+	{
+		if (str1[i] > str2[i])
+			return -1;
+
+		if (str1[i] < str2[i])
+			return 1;
+	}
+
+	return 0;
+}
+
+void strCopy(char* direction, const char* toCopy)
+{
+	size_t toCopyLen = strLength(toCopy);
+
+	for (size_t i = 0; i < toCopyLen; i++)
+		direction[i] = toCopy[i];
+
+	direction[toCopyLen] = '\0';
+}
+
+void swapOrders(PizzaOrder& order1, PizzaOrder& order2)
+{
+	PizzaOrder temp = order1;
+	order1 = order2;
+	order2 = temp;
+}
+
+void sortOrders(PizzaOrder* orders, const size_t ordersSize)
+{
+	for (size_t i = 0; i < ordersSize; i++)
+	{
+		size_t minIndex = i;
+
+		for (size_t j = i + 1; j < ordersSize; j++)
+			if (strCompare(orders[minIndex].address, orders[j].address) == -1)
+				minIndex = j;
+
+		swapOrders(orders[minIndex], orders[i]);
+	}
+}
+
+void getOrdersFromBinary(PizzaOrder*& orders, size_t& ordersSize)
+{
+	std::ifstream in(FILE_PATH, std::ios::binary);
+
+	if (!in.is_open())
+	{
+		std::cerr << "File couldn't open";
+		exit(0);
+	}
+
+	size_t fileSize = getFileSize(in);
+	ordersSize = fileSize / sizeof(PizzaOrder);
+
+	orders = new PizzaOrder[ordersSize + 1];
```

Comment:
why +1

Reply:
because I wanted to have 1 free space in the array for the order I read from the console

Reply:
but you use this function when you print orders.
Then you have one more order printed

Reply:
Redesign your solution. Your functions have unexpected behaviors. It's not ok a function called getOrdersFromBinary to return one empty object at the end of the result array.